Migrated to PHP 7 and added tests.

DrDub opened the pull request:
The zval change impacted this code throughout, this close to a rewrite.

I know this is very old code, but I'm looking at whether it can help speed up the tensor library in Rubix/ML. A contiguous C array of floats is exactly what we need to interface to the linear algebra C libraries.

DrDub committed:
Migrated to PHP 7 and added tests.

The zval change impacted this code throughout.

nikic commented on buffer.c:
```diff
@@ -124,63 +61,41 @@ zend_class_entry *double_array_ce;
-static void array_buffer_free_object_storage(buffer_object *intern TSRMLS_DC)
-{
-	zend_object_std_dtor(&intern->std TSRMLS_CC);
+	buffer_object * intern = php_buffer_fetch_object(object);
```
```suggestion
	buffer_object *intern = php_buffer_fetch_object(object);
```
Here and elsewhere

nikic commented on buffer.c:
```diff
@@ -124,63 +61,41 @@ zend_class_entry *double_array_ce;
 	}
 
+	zend_object_std_dtor(&intern->std);
 	efree(intern);
```
Should be dropped, this will cause a double free.

DrDub replied:
Sorry about that. Too bad it wasn't caught from the texts.

nikic commented on buffer.c:
```diff
@@ -124,63 +61,41 @@ zend_class_entry *double_array_ce;
 
-	buffer_object *intern = emalloc(sizeof(buffer_object));
-	memset(intern, 0, sizeof(buffer_object));
+	buffer_object *intern = emalloc(sizeof(buffer_object) +
```
Prefer zend_object_alloc

nikic commented on buffer.c:
```diff
@@ -199,18 +114,18 @@ PHP_METHOD(ArrayBuffer, __construct)
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l", &length) == FAILURE) {
-		zend_restore_error_handling(&error_handling TSRMLS_CC);
+	zend_replace_error_handling(EH_THROW, NULL, &error_handling);
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &length) == FAILURE) {
```
Use zend_parse_parameters_throw instead of replacing error handling.

DrDub replied:
I changed the other cases where it stopped error handling but there are other `zend_parse_parameters` that just return on failure. (I looked at the header file and `zend_parse_parameters` and `zend_parse_parameters_throw` are currently equivalent but maybe you want to signal clearly all errors in parameter parsing throw exceptions.)

nikic replied:
zend_parse_parameters_throw is the same as zend_parse_parameters in PHP 8, because throw is not the only option. Previously zend_parse_parameters_throw was required to throw on invalid arguments.

nikic commented on buffer.c:
```diff
@@ -199,18 +114,18 @@ PHP_METHOD(ArrayBuffer, __construct)
 	long length;
```
long -> zend_long

nikic commented on buffer.c:
```diff
@@ -224,55 +139,59 @@ PHP_METHOD(ArrayBuffer, serialize)
 	/* Serialize buffer as string */
-	ZVAL_STRINGL(zv_ptr, (char *) intern->buffer, (int) intern->length, 0);
-	php_var_serialize(&buf, &zv_ptr, &var_hash TSRMLS_CC);
+        zend_string * zstr = zend_string_init((char *) intern->buffer, (int) intern->length, 0);
```
int -> size_t

nikic commented on buffer.c:
```diff
@@ -224,55 +139,59 @@ PHP_METHOD(ArrayBuffer, serialize)
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &str, &str_len) == FAILURE) {
+		return;
+	}
+	if (str_len == 0) {
```
Hm, why do we need this check?

DrDub replied:
Looking at `zend_API.h`, I think it might be because this is asking for a pointer to a char with `"s"` the code will return a NULL for `str` when `str_len==0`:

```c
static zend_always_inline bool zend_parse_arg_string(zval *arg, char **dest, size_t *dest_len, bool check_null, uint32_t arg_num)
{
        zend_string *str;

        if (!zend_parse_arg_str(arg, &str, check_null, arg_num)) {
                return 0;
        }
        if (check_null && UNEXPECTED(!str)) {
                *dest = NULL;
                *dest_len = 0;
        } else {
```

That NULL pointer might choke the call

```c
	if (!php_var_unserialize(zbuf, &p, max, &var_hash)) {
```

(Not familiar how `php_var_unserialize` will handle `p` pointing to variable pointing to NULL.

nikic replied:
I believe str can only be null if `s!` is used (i.e a nullable parameter). With `s` it should be always non-null. For str_len == 0, it would just be an empty string.

DrDub replied:
Oh, yes, I missed the `check_null` part.

Removed the check.

nikic commented on buffer.c:
```diff
@@ -282,56 +201,67 @@ PHP_METHOD(ArrayBuffer, unserialize)
 	}
 
+	zend_object_std_dtor(&intern->std);
 	efree(intern);
```
efree needs to be dropped.

DrDub replied:
Thanks. The free'ing and the destructor gave me a lot of grief but I had hoped the tests should have caught this up. I wonder what type of testing would have found this.

nikic replied:
Either valgrind or address sanitizer can be used to catch such issues. Valgrind is slow but probably easier for extensions. (Unless doing an in-tree build, in which case `--enable-address-sanitizer` can be used.)

nikic commented on buffer.c:
```diff
@@ -282,56 +201,67 @@ PHP_METHOD(ArrayBuffer, unserialize)
-	if (intern->buffer_zval) {
-		zval_ptr_dtor(&intern->buffer_zval);
+	buffer_view_object *intern = php_buffer_view_fetch_object(obj);
+	if(!intern){
```
I don't think `!intern` is possible here.

nikic commented on buffer.c:
```diff
@@ -282,56 +201,67 @@ PHP_METHOD(ArrayBuffer, unserialize)
-	zend_object_std_dtor(&intern->std TSRMLS_CC);
-
-	if (intern->buffer_zval) {
-		zval_ptr_dtor(&intern->buffer_zval);
```
Don't we still need this in some form?

DrDub replied:
I removed it as it was part of the `zend_objects_store` functionality that got removed in PHP 7.

Maybe the `array_buffer_view_free` is missing this line at the end?

```c
	if (!Z_ISUNDEF(intern->buffer_zval)) {
		zval_ptr_dtor(&intern->buffer_zval);
	}
```

nikic commented on buffer.c:
```diff
@@ -454,25 +371,20 @@ zval *buffer_view_offset_get(buffer_view_object *intern, size_t offset)
 		zval_ptr_dtor(&value);
 	} else {
-		long lval = get_long_from_zval(value);
+		long lval = zval_get_long(value);
```
zend_long

DrDub replied:
Yes. I also changed other variables receiving the output of `zval_get_long` to `zend_long`.

nikic commented on buffer.c:
```diff
@@ -559,151 +468,160 @@ static int array_buffer_view_has_dimension(zval *object, zval *zv_offset, int ch
+		BUFFER_SWITCH_CASE(int32)
+		BUFFER_SWITCH_CASE(uint32)
+		BUFFER_SWITCH_CASE(float)
+		BUFFER_SWITCH_CASE(double)
```
Why do we need this? These are all the same pointer.

DrDub replied:
The pointer is the same yes, but the size to copy is expressed in the length of the type. 

I will streamline the code  more elegantly by introducing a temporary variable `type_size`.

DrDub replied:
My bad, there was a `buffer_view_get_bytes_per_element` function just for that. Using it made the code much better.

nikic commented on buffer.c:
```diff
@@ -1052,18 +987,52 @@ PHP_MINIT_FUNCTION(buffer)
+    php_info_print_table_row(2, "buffer module version", BUFFER_VERSION);
+    php_info_print_table_row(2, "author", "nikic");
+    php_info_print_table_end();
+    DISPLAY_INI_ENTRIES();
```
Wrong indentation here.

DrDub commented:
Thanks so much for your thorough review. 

I realize the initial code in the PR did not pass all the tests under a version of PHP compiled without debug enabled. I'll see if fixing the issues you mention helps otherwise I might ask for more help. 

Any further comments will be appreciated.

DrDub committed:
Changes from review

DrDub committed:
iterator destructor fix

DrDub commented:
OK, all tests now pass both with a php binary with debug enabled or disabled.

Together with the changes you identified, I added the destructor call for the buffer zval in `array_buffer_view_free` and fixed some bugs in `array_buffer_view_read_dimension` and `buffer_view_iterator_dtor`.

The code now compiles without warnings, which I should have heed before as they pointed to the `array_buffer_view_read_dimension` bug.

nikic commented on buffer.c:
```diff
@@ -224,114 +133,129 @@ PHP_METHOD(ArrayBuffer, serialize)
-
-	zend_object_std_init(&intern->std, class_type TSRMLS_CC);
-	object_properties_init(&intern->std, class_type);
+	buffer_view_object *intern = emalloc(sizeof(buffer_view_object) +
+			                                         zend_object_properties_size(class_type));
```
```suggestion
	buffer_view_object *intern = zend_object_alloc(sizeof(buffer_view_object), class_type);
```

nikic commented on buffer.c:
```diff
@@ -361,42 +285,32 @@ zend_object_value array_buffer_view_create_object(zend_class_entry *class_type T
 	new_object->buffer_zval = old_object->buffer_zval;
-	if (new_object->buffer_zval) {
-		Z_ADDREF_P(new_object->buffer_zval);
+	if (!Z_ISUNDEF(new_object->buffer_zval)) {
```
Can omit the Z_ISUNDEF check, it's fine to copy an UNDEF zval.

nikic commented on buffer.c:
```diff
@@ -454,25 +365,20 @@ zval *buffer_view_offset_get(buffer_view_object *intern, size_t offset)
 		}
 
-		zval_ptr_dtor(&value);
+		zval_ptr_dtor(value);
```
This probably needs to be removed, this is now destroying the offset argument, not the conversion result.

nikic commented on buffer.c:
```diff
@@ -494,63 +400,61 @@ void buffer_view_offset_set(buffer_view_object *intern, long offset, zval *value
 }
 
-static zval *array_buffer_view_read_dimension(zval *object, zval *zv_offset, int type TSRMLS_DC)
+static zval* array_buffer_view_read_dimension(zval *object, zval *zv_offset, int type, zval *retval)
```
```suggestion
static zval *array_buffer_view_read_dimension(zval *object, zval *zv_offset, int type, zval *retval)
```

nikic commented on buffer.c:
```diff
@@ -454,25 +365,20 @@ zval *buffer_view_offset_get(buffer_view_object *intern, size_t offset)
-	return retval;
 }
 
 void buffer_view_offset_set(buffer_view_object *intern, long offset, zval *value)
```
```suggestion
void buffer_view_offset_set(buffer_view_object *intern, zend_long offset, zval *value)
```

nikic commented on buffer.c:
```diff
@@ -559,151 +463,144 @@ static int array_buffer_view_has_dimension(zval *object, zval *zv_offset, int ch
-		retval = zend_is_true(value);
-		zval_ptr_dtor(&value);
+		zval value;
+			    buffer_view_offset_get(intern, offset, &value);
```
Weird indent

nikic commented on buffer.c:
```diff
@@ -559,151 +463,144 @@ static int array_buffer_view_has_dimension(zval *object, zval *zv_offset, int ch
-		zval_ptr_dtor(&value);
+		zval value;
+			    buffer_view_offset_get(intern, offset, &value);
+		retval = zend_is_true(&value);
```
Could directly return here, without the extra variable.

The dtor here was dropped because it's always an int/double, so no dtor necessary?

DrDub replied:
(Removed the extra variable)

Yes, the dtor was dropped because `value` lives in the stack, not in the heap.

nikic commented on buffer.c:
```diff
@@ -835,29 +741,29 @@ PHP_FUNCTION(array_buffer_view_wakeup)
 
-	retval = buffer_view_offset_get(intern, offset);
-	RETURN_ZVAL(retval, 1, 1);
+	buffer_view_offset_get(intern, offset, &retval);
```
Can directory write into `return_value` here.

DrDub replied:
Sorry, I don't understand what you mean. Write directly into `return_value` how?

This is the current code:

```c
	buffer_view_offset_get(intern, offset, &retval);
	RETURN_ZVAL(&retval, 1, 0);
```

I gather the `RETURN_ZVAL` can be elided in some way...?

nikic replied:
It's possible to write `buffer_view_offset_get(intern, offset, return_value);`

DrDub commented on php_buffer.h:
```diff
@@ -1,42 +1,90 @@
+	size_t length;
+
+	/* For Iterator methods */
+	size_t current_offset;
```
After this, I want to do a migration to PHP 8, maybe leave that change for then? Otherwise I can give this a try now, just let me know.

Something else we need for Rubix is for Buffer to implement the Countable interface, that's also next for me but I hope the current migration is solid before adding changes.

DrDub committed:
Update buffer.c

Co-authored-by: Nikita Popov <github@npopov.com>

DrDub committed:
Update buffer.c

Co-authored-by: Nikita Popov <github@npopov.com>

DrDub committed:
Update buffer.c

Co-authored-by: Nikita Popov <github@npopov.com>

DrDub committed:
Update buffer.c

Co-authored-by: Nikita Popov <github@npopov.com>

DrDub commented on buffer.c:
```diff
@@ -224,114 +133,129 @@ PHP_METHOD(ArrayBuffer, serialize)
-	}
+        }
+        if(Z_TYPE_P(zbuf) != IS_STRING) {
+		zend_throw_exception(NULL, "Could not unserialize buffer: not a string", 0);
```
Good point.

Looking at the code in `ext/standard/var_unserializer.c`, this zbuf is a pointer inside the var hash which in turn has its destructors called in the `var_destroy` function (chained from the `PHP_VAR_UNSERIALIZE_DESTROY` on the exit). So it should be fine.

DrDub committed:
More changes from review

DrDub commented:
I added the latest changes. Slowly gaining on my understanding of PHP internals. Thanks so much for your guidance.

DrDub committed:
returning directly

nikic commented:
Thanks!

It would probably be good to add some basic Github Actions setup to check that things build with versions which are intended to be supported. Always a bit tricky with differences across versions.
